File: 216-combination-sum-iii/216-combination-sum-iii.cpp
class Solution {
public:
    vector<vector<int>> combinationSum3(int k, int n) {
        vector<vector<int>> result;
        vector<int> current;
        backtrack(result, current, n, k, 1);
        return result;
    }
    void backtrack(vector<vector<int>> &result, vector<int> current, int n, int k, int ind)     {
        if(current.size() == k){
            if(n == 0){
                // vector<int> copy = current;
                result.push_back(current);
            }
            return;
        }    
        for(int i=ind; i<=9; i++){
            current.push_back(i);
            backtrack(result, current, n-i, k, i+1);
            current.pop_back();
        }
    }
};

File: 216-combination-sum-iii/216-combination-sum-iii.java
class Solution {
    public List<List<Integer>> combinationSum3(int k, int n) {
        List<List<Integer>> result = new ArrayList<>();
        backtrack(result, new ArrayList<>(), k, n, 1);
        return result;
    }
    private void backtrack(List<List<Integer>> result, List<Integer> current, int k, int n, int index)     {
        if(current.size() == k){
            if(n == 0){
                result.add(new ArrayList<>(current));
            }
            return;
        }    
        for(int i=index; i<=9; i++){
            current.add(i);
            backtrack(result, current, k, n-i, i+1);
            current.remove(current.size()-1);
        }
    }
}